### Title: NfeKO RNAseq
### Author: Kate Johnson
### Date: 12/31/2024
### Add'tl Notes:

# setup
libraries
```{r, warning=FALSE}
# all-purpose data analysis tools
library(tidyverse)
# ggplot extension
library(cowplot)
# repeling labels in ggplot
library(ggrepel)
#correlation plots
library(corrplot)
# easier string formatting
library(glue)
# filepath manipulation
library(fs)

# edgeR and limma for DE analysis
library(edgeR)
library(limma)

# for reading/writing .gct files
#library(cmapR)

# for working with matrices (especially sparse ones)
library(Matrix)

#complex heatmap /plotting
library(ComplexHeatmap)
library(circlize) #for colorRamp2 color scaling
library(monochromeR) # for generating palettes
library(ggpubr) #arranging plots, stat_cor()
library(emojifont) #for help rendering unicode text

#GSEA
library(clusterProfiler)
library(enrichplot)

# SET THE DESIRED ORGANISM HERE
organism = "org.Mm.eg.db"
#BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)
```
theme
```{r}
# Plot settings ----
#compress PDFs
tools::compactPDF('mypdfs/',  gs_quality='screen')
###PLOTTING THEME ###
theme_update(text = element_text(size = rel(4)),
             legend.title = element_text(size = rel(2.5)),
             legend.key.size = unit(.1, 'inches'),
             legend.text = element_text(size = rel(2)),
             panel.background = element_rect(fill = "white", colour = "black"),
             axis.line = element_blank(),
             axis.text.x = element_text(size=rel(3)),
             axis.text.y = element_text(size=rel(3)),
             axis.title.y = element_text(margin=margin(t = 0.5, r = 0.5, 
                                                       b = 0.5, l = 0.5), size=rel(3)),
             axis.title.x = element_text(size=rel(3)),
             plot.title = element_text(size = rel(3), face= "bold", hjust=0.5),
             plot.margin = unit(c(0.08, 0.1, 0.01, 0.01), "inches"),
             strip.text.x = element_text(size = rel(3), face = "bold", margin = margin(0.05,0.05,0.05,0.05, "cm")),
             strip.text.y = element_text(size = rel(3), face = "bold", margin = margin(0.05,0.05,0.05,0.05, "cm")),
             strip.background = element_rect(linewidth=0.35, color="black"),
             legend.key = element_rect(colour = NA, fill = NA)) #trbl
```


directories
```{r}
wd = getwd()
# For reading in counts/metadata
input_dir = fs::path(wd, "../input/data/2_formatted_data/bcbio_genome")
# For saving plots  
output_dir = fs::path(wd, "../output/bcbio_genome")

```

sources
```{r}

```


# Read in Data
```{r, warnings =F}
#counts
counts = read_csv(fs::path(input_dir, "20241231_Nfekoseq_counts.csv")) 
#rename gene column
colnames(counts)[1] = "gene"

# format matrix to be ready for de analysis
genes <- counts$gene
counts$gene <- NULL
rownames(counts) <- make.unique(genes) #deal with duplicate gene names
                    #mostly this should be resolved after filtering

#meta data
metadata = read_csv(fs::path(input_dir, "20241231_Nfekoseq__metadata.csv")) %>%
           #drop unused first col
           dplyr::select(-1) %>%
           #refactor relevant columns
           mutate(Target = factor(Target, levels = c("sgCtrl", "sgNfe2l1","sgNfe2l2", "doubleKO")),
                  Low_Reads = ifelse(Sample_Name %in% c("sgCtrl_g1_rep1", "sgCtrl_g2_rep1", "sgNfe2l1_g2_rep1"),T,F),
                  Target_guide = str_c(Target, "_", Guide)) #%>%
           #filter(Low_Reads == F)#%>%
           #filter(Sample_Name != "sgCtrl_g1_rep1")

#remove plus signs from colnames to avoid subsequent issues
counts_filt = counts[,metadata$Sample_Name]
rownames(counts_filt) <- make.unique(genes)
 
```
# QC
Data Normalization
```{r}
#set cpm thresholds (keep low to start)
cpm_threshold = 10
cpm_needed = 2

groups <- factor(metadata$Target)
dge = DGEList(counts = counts_filt, group = groups)
dge_norm = calcNormFactors(dge)
#require cpm>10 in at least 2 samples 
isexpr = (rowSums(cpm(dge_norm) > cpm_threshold) >= cpm_needed)
  
dge_norm_fltd = dge_norm[isexpr, ]

cpm = cpm(dge_norm_fltd) #dropped to ~9.2k for cpms10
cpm_all = cpm(dge_norm) #~54k reads total

#save cpm heatmaps for plotting in morpheus
#write.csv(as.matrix(cpm), fs::path(output_dir, "cpm_tables/ctrlg1r1_dropped_only/20241231_Nfekoseq_cpmTable_filt_cpms10_2samples.csv"))
#write.csv(as.matrix(cpm_all), fs::path(output_dir, "cpm_tables/ctrlg1r1_dropped_only/20241231_Nfekoseq_cpmTable_all.csv"))

#consider tossing low read count samples (sgCtrl g1 rep1, sgNfe2l1 g1 rep2 etc) before proceeding with DE analysis

```

PCA
```{r}
# run pca
pca = prcomp(t(cpm), center=T, scale=T)
# combine pca results with metadata for easy plotting
pca_res = bind_cols(metadata, as_tibble(pca$x)) 

# plot
# color by treatment
p1<-ggplot(pca_res, aes(PC1, PC2, fill=Target)) + 
    geom_point(size=3, shape =21, color = "black") +
    #scale_fill_manual(values = c("grey", "#F68064")) +
    theme(text = element_text(size=6),
          legend.position = "top") +
    guides(fill=guide_legend(nrow=2))
p1
# color by timepoint
p2<-ggplot(pca_res, aes(PC1, PC2, fill=Guide)) + 
    geom_point(size=3, shape = 21, color = "black") +
    #scale_fill_manual(values = c("grey95",  "grey55","grey20"))+
    theme(text = element_text(size=6),
          legend.position = "top")
p2
# color by replicate
p3<-ggplot(pca_res, aes(PC1, PC2, fill=Replicate)) + 
    geom_point(size=3, shape =21, color = "black") +
    theme(text = element_text(size=6),
          legend.position = "top")
p3

p4<-ggplot(pca_res, aes(PC1, PC2, fill=Low_Reads)) + 
    geom_point(size=3, shape =21, color = "black") +
    theme(text = element_text(size=6),
          legend.position = "top")
p4

grid <-ggarrange(p1,p2,p3,p4, nrow=1)

grid
# save plots
#save_plot(fs::path(output_dir, "QC/PCA/ctrlg1r1_dropped_only/cpms10_PCA.pdf"),
       #   grid,dpi=90, base_aspect_ratio =3.75)


```


replicate correlation
```{r}
# commenting out for now bc not particularly informative
#pairwise correlations between samples for each treatment
for(treatment in unique(metadata$Target_guide)){
    suppressWarnings(
        gplots::heatmap.2(cor(counts[, metadata$Target_guide == treatment]),
                          trace = 'none', Rowv=F, Colv=F, 
                         col=viridis::viridis(75),
                          margins = c(10, 10))
    )
}

# table of all pairwise correlations < 0.9
# get all pairwise correlations
cormelt = as_tibble(reshape2::melt(cor(counts)))
colnames(cormelt) = c('s1', 's2', 'cor')

filt_cormelt = cormelt %>%
    # extract pieces of name for sample 1
    extract(s1, c('target', 'guide', 'replicate1'), '(.*)_(.*)_(.*)') %>%
    # extract pieces of name for sample 2
    extract(s2, c('target2', 'guide2', 'replicate2'), '(.*)_(.*)_(.*)') %>%
    # only keep samples where treatment and pool match
    # replicate1 < replicate2 removes duplicate comparisons
    filter(target == target2, guide == guide2, replicate1 < replicate2) %>%
    # comparison column makes for easier reading
    mutate(comparison = str_c(replicate1, '-', replicate2),
           target = factor(target, levels = c("sgCtrl", "sgNfe2l1", "sgNfe2l2", "doubleKO"))) %>%
    # remove extraneous columns
    dplyr::select(-target2, -guide2, -replicate1, -replicate2) #%>%
    # all correlations are above 0.9 so keep everything for now
    # filter(cor < .9)


p<-ggplot(filt_cormelt,
       aes(x = target, y = guide, fill = cor)) +
       geom_tile() +
       geom_text(aes(label = round(cor,3)), 
                  hjust = 0.5, vjust = 0.5) +
       facet_wrap(~comparison) +
       theme(axis.text.x = element_text(angle =45, hjust =1))
p  
#save plot 
save_plot(fs::path(output_dir, "QC/20241231_replicate_correlation_all_samples.pdf"),p, dpi = 90, base_aspect_ratio = 3)
```


#DE analysis# DE analysis
The general steps for modeling are: 
1. Filter cpm and metadata based on the analysis at hand
2. Make a design matrix specifying the experimental condition of each sample 
3. Run `voom()` to transform and prepare data
4. Run initial `lmFit()` to fit data
5. Setup contrasts between groups for tests you want to run
    1. List contrasts
    2. Make contrast matrix
6. Run `contrast.fit()` to fit contrasts
7. Run `eBayes` to estimate statistic results
8. Gather results into useful tables

```{r}
# repeat dge list creation here in case filtering is necessary of counts table after QC
  #for now including all samples so this is redundant
# set cpm thresholds (keep low to start)
cpm_threshold = 10 #X
cpm_needed = 2 #Y


## re-do dge list object creation using adjusted counts table
groups <- factor(metadata$Target)
dge = DGEList(counts = counts_filt, group = groups)
dge_norm = calcNormFactors(dge)

#require cpm>X in at least Y samples
isexpr = rowSums(cpm(dge_norm) > cpm_threshold) >= cpm_needed
dge_norm_fltd = dge_norm[isexpr, ]


# DESIGN MATRIX----
# create factor with all conditions in experiment
f <- factor(groups, levels=unique(groups))

# build a design matrix for modeling with limma voom and indicating condition for each sample
design <- model.matrix(~0 + f)

# renames column names in the design (remove extra "f" at beginning of each group name)
colnames(design) <- sub("f", "", colnames(design))
rownames(design) <- colnames(counts_filt)

##LIMMA VOOM 
# limma voom transforms count data to log2-counts per million (log2-cpm) with associated weights
# estimate the mean-variance relationship and use this to compute appropriate observational-level weights.
# The data are then ready for linear modeling.
y <- voom(dge_norm_fltd, design, plot=T)

# fit linear model for each gene given a series of arrays
fit <- lmFit(y, design)

##CONTRAST MATRIX----
# to start compute all contrasts as each target relative to SCR control
# 1. get all targets that arent control
# 2. subtract each sample from its time matched control
all_cond = levels(groups)[!grepl("sgCtrl", (levels(groups)))]
#time/celltype match each condition to control
contrasts_list = str_c(all_cond, " - sgCtrl")

contrast.matrix <- makeContrasts(contrasts = contrasts_list, levels = design)

#do some renaming of scores to facilitate pivoting later on and make our lives easier...
contrast.matrix = contrast.matrix %>%
                  as.data.frame %>% #convert to df for now for manipulation
                  rename_with(., ~gsub(" - .*", "", .)) %>%
                  as.matrix()

fit.cont <- contrasts.fit(fit, contrast.matrix)
e.fit <- eBayes(fit.cont)

#first bind together all LFC lists for all genes from ea. contrast
log2FC.all = as.data.frame(map_dfc(1:length(colnames(e.fit$contrasts)),
                             function(coef) topTable(e.fit, coef=coef, number=Inf, adjust = "BH",
                                                     sort.by='none')$logFC)) %>%
      data.frame()

adj_p = as.data.frame(map_dfc(1:length(colnames(e.fit$contrasts)),
                                    function(coef) topTable(e.fit, coef=coef, adjust = "BH",
                                                            number=Inf,
                                                            sort.by='none')$adj.P.Val))

rownames(log2FC.all) = rownames(adj_p) = make.unique(rownames(dge_norm_fltd), sep = ".")
colnames(log2FC.all) = colnames(adj_p) = colnames(e.fit$contrasts)

#turn log2FC.all.binary into matrix of -1,0,1
FDR_thresh = 0.05
log2FC.all.binary = adj_p %>%
              mutate_all(~ifelse(. < FDR_thresh, 1, 0))

#filter LFC matrix accordingly
log2FC.all.binary[(log2FC.all < 0) == T] = -1 * log2FC.all.binary[(log2FC.all < 0) == T]
names = rownames(log2FC.all.binary)
log2FC.all.binary = log2FC.all.binary %>%
                    data.frame()

DE_log2FC.all = log2FC.all[rowSums(log2FC.all.binary != 0) > 0,] %>%
          data.frame()
rownames(DE_log2FC.all) = rownames(log2FC.all)[rowSums(log2FC.all.binary != 0) > 0]

#filter DE matrix based on LFC thresh
LFC_thresh = 0.5
DE_LFC_thresh = DE_log2FC.all %>%
          filter_all(any_vars(abs(.) > LFC_thresh))

#zero non-sig genes (?)
#DE_zeroed = DE_LFC_thresh %>%
 #           mutate_all(~ifelse(is.na(.), 0,.)) 

#save tables
write.csv(as.matrix(DE_LFC_thresh), fs::path(output_dir,                               "DE_analysis/FDR0.05_LFC0.5/LFC_matrix.csv"))

write.csv(as.matrix(log2FC.all.binary),
          fs::path(output_dir,                         "DE_analysis/FDR0.05_LFC0.5/20241231KJ_FDR0.05_LFC0.5_cpm10_DE-genes-binary-matrix.csv"))
```

#single gene plots of glu metabolism genes
```{r}
#format filtered cpms
temp = cbind(gene = rownames(cpm), cpm) 

plot_me = temp %>%
          data.frame() %>%
          pivot_longer(cols = -gene, names_to = "Sample_Name", values_to = "cpm") %>%
          mutate(cpm = as.numeric(cpm)) %>%
          separate(Sample_Name, into = c("Target", "Guide", "Replicate")) %>%
          group_by(gene, Target, Guide) %>%
          summarise(n = n(),
                   avg_cpm = mean(cpm, na.rm=T),
                   se = sd(cpm)/sqrt(n)
          ) %>%
          mutate(Target = factor(Target, levels = c("sgCtrl", "sgNfe2l1", "sgNfe2l2", "doubleKO"))) %>%
          arrange(Target)

glu_pathway = read_csv(fs::path(wd,"../../../Interaction_Scoring/input/other/Glutamate_metabolic_pathway.csv")) %>%
              mutate(Gene = str_to_title(Gene))
my_genes = c(glu_pathway$Gene, "Slc7a11","Slc25a22", "Slc25a12", "Slc25a13", "Ndufs3", "Atp5f1", "Nfe2l1", "Nfe2l2")
up <- c("Ears2",'Gls', "Gfpt1", "Glud1", "Got2")
down <- c("Eprs", "Glul", "Gclc", "Gclm", "Gss","Got1")

plot_me_filt = plot_me %>%
               filter(grepl("Cul3|Keap1|Ddi2", gene)) %>%
               mutate(Target_Guide = str_c(Target, "_", Guide),
                      Target_Guide = factor(Target_Guide, levels = unique(Target_Guide)),
                      gene = factor(gene, levels = c("Ddi2", "Cul3", "Keap1")))

p<-ggplot(plot_me_filt, aes(x = Target_Guide, y = avg_cpm, fill = Target, alpha = Guide)) +
  scale_fill_manual(values = c("grey90", "#7CAE00", "#00BFC4", "grey30")) +
  scale_alpha_manual(values = c(1,0.5)) +
  geom_col(position = "dodge", color = "black") +
  geom_errorbar(mapping = aes(ymin = avg_cpm -se, ymax = avg_cpm+se), color = "black", alpha = 1) +
  facet_wrap(~gene, scales = "free_y") +
  theme(axis.text.x = element_text(angle = 45, hjust=1, size = 6))
p
save_plot(fs::path(output_dir, "single_gene_plots/Glu_TF_regulators_Cul3-Ddi2-Keap1.pdf"), p , base_aspect_ratio = 2.5)
```


#Number of DEGs plot
```{r}
#plot DEGs per sample
#code is slow -- find way to do this better?
stats = log2FC.all.binary %>%
        #rowwise() %>% #try using reframe ..?
        reframe(up = colSums(.>0),
                down = colSums(.<0)) %>%
        mutate(Sample = colnames(log2FC.all.binary)) %>%
        left_join(metadata %>% select(-c(Replicate, Sample_Rep, 
                                         Barcode, Column, Row, Well, 
                                         Cell_type_barcode)), by = "Sample") %>%
        pivot_longer(cols = c("up", "down"), names_to = "DEG", values_to = "count") %>%
        mutate(Treatment_drug = str_c(Treatment, "_", Drug_abbrev),
                #collapse DMSO info into Concentration info
                #Concentration = ifelse(grepl("DMSO", Treatment_drug), "DMSO",
                # as.character(Concentration)),
                      Concentration = factor(Concentration, levels = c("None", "Lo", "Hi")),
                      #strip DMSO info from drug names
                      Treatment_drug = gsub("LSG_DMSO", "LSG_None", 
                                        gsub("[+]DMSO", "", Treatment_drug)),
                      #force order of drugs
                      Treatment_drug = factor(Treatment_drug, 
                                          levels = c("Control_None", "LSG_None",
                                                     "LSG_A", "LSG_E", "LSG_C", "LSG_D",
                                                     "LSG_A+C", "LSG_A+D", "LSG_E+C", "LSG_E+D",
                                                     "LSG_A+E", "LSG_D+C")),
                      #Drug_short = gsub("[+]DMSO", "", gsub("^DMSO$","None",Drug_short)),
                      Drug_type = ifelse(Drug_type == "Single" | Drug_type == "None",
                                         gsub("[+]DMSO", "", gsub("^DMSO$","None",Drug_short)), 
                                    ifelse(Drug_abbrev %in% c("A+E", "D+C"), 
                                           "Pair \n(Single Target)", 
                                           "Pair \n(Dual Target)")),
                      Drug_type = factor(Drug_type, 
                                         levels = c("None", "AAA", "Erastin", 
                                                    "Cerestat", "D-AP5",
                                                     "Pair \n(Dual Target)", 
                                                    "Pair \n(Single Target)"))) %>%
                filter(Cell_type == "Tcell") %>%
                arrange(Cell_type, Treatment_drug, DMSO, Concentration ,desc(Time),count) %>%
                mutate(Drug_short_conc = factor(Drug_short_conc, levels = unique(Drug_short_conc))) %>%
                unique()

p<-ggplot(stats, aes(x = Drug_short_conc, y= count, fill = Treatment_drug, alpha = Concentration)) +
  geom_col(data = stats %>% filter(DEG == "up"), 
           color = "black", position = position_dodge2(preserve = "single", width = 1)) +
  geom_col(data = stats %>% filter(DEG == "down"), aes(y = -count), 
           position = position_dodge2(preserve = "single", width = 1), 
           color = "black") +
  #scale_alpha_manual(values = c(1,0.5,1)) +
  scale_fill_manual(values = c("Control_None" = "grey", "LSG_None" = "grey",
                                     "LSG_A" = "#A6DFF7", "LSG_E" = "#B9CEE6",
                                     "LSG_C" = "#8C9FCA", "LSG_D" = "#6D91CB",
                                     "LSG_A+C" = "#5BABA4","LSG_A+D" = "#68C3A5",
                                     "LSG_E+C" ="#539C84", "LSG_E+D" ="#3E7563",
                                     "LSG_A+E" ="#9E7FB8", "LSG_D+C" ="#5E4C6E",
                                     "#D278AE", "#871F78", "#C0ACD0")) +
  facet_grid(Time~Drug_type, scales = "free") +
  #ggh4x::facet_grid2(Time~Drug_type, independent = "y", scales = "free") +
  labs(x = NULL, y = "Number DEGs", title = "Tcells") +
  theme(axis.text.x = element_text(angle=45, hjust=1, size = rel(2.7)),
        strip.text = element_text(size=rel(0.6)),
        legend.position = "none") +
      guides(fill=guide_legend(nrow=5))

#saveplot
save_plot(path(output_dir, "DE_analysis/LSG_NoDrug/FDR0.05_LFC1/DEGs_TCells.pdf"), p, 
          base_aspect_ratio =1.6)

#save tables
write.csv(as.matrix(BMDC_T_DE_merge), path(output_dir,                               "DE_analysis/LSG_NoDrug/FDR0.05_LFC1/DCT_merge_LFC_matrix.csv"))
write.csv(as.matrix(log2FC.all.binary),
          path(output_dir,                         "DE_analysis/LSG_NoDrug/FDR0.05_LFC1/20241003KJ_FDR0.05_LFC1_cpm20_DE-genes-binary-matrix.csv"))
```


#complex heatmap
#complex heatmap plot fxn
move to plotting functions!
```{r}
do_plot = function(file_name, w, h, p) {
  pdf(file=fs::path(output_dir, file_name), width = w, height = h)
      draw(p)
      dev.off()
}
```

#making heatmap
```{r}
#set heatmap params----
#set seed for kmeans
set.seed(123)

#set fast clustering
ht_opt$fast_hclust = TRUE
ht_opt$message = FALSE


#format df for complex heatmap
df = BMDC_T_DE_merge %>%
     column_to_rownames("gene") %>%
     as.matrix()

#prepare annotations
top_annot = data.frame(Sample = colnames(df)) %>%
            separate(Sample, into = c("Treatment", "Drug", "Concentration",
                                    "Empty", "Cell_type","Time"), 
                     sep ="_", remove =F) %>%
            inner_join(metadata %>% 
                      select(c(Sample, Drug_short, Concentration)) %>%
                      unique(),
                      by = c("Sample", "Concentration")) %>%
            select(c(Sample, Cell_type, Time, Drug_short, Concentration)) %>%
            mutate(Concentration = ifelse(grepl("[+]DMSO", Drug_short), "Hi (+DMSO)", Concentration),
                   Concentration = factor(Concentration, 
                                          levels = c("None", "Lo", "Hi (+DMSO)", "Hi")),
                   Drug_short = gsub("[+]DMSO", "", Drug_short),
                   Drug_short = factor(Drug_short,
                                levels = c(
                                "DMSO",
                                "D-AP5",
                                "AAA" ,
                                "Erastin" ,
                                "Cerestat" , 
                                "D-AP5+Cerestat", "AAA+Erastin",
                                "AAA+D-AP5","Erastin+D-AP5",
                                "AAA+Cerestat", "Erastin+Cerestat")),
                   Time = factor(Time, levels = c("4h", "24h", "72h"))) %>%
            arrange(Cell_type, Time, Drug_short, Concentration) %>%
            rename("Drug" = "Drug_short") %>%
            column_to_rownames("Sample") %>%
            filter(!(Concentration %in% c("Lo", "Hi (+DMSO)"))) %>%
            mutate(CellType_Time = str_c(Cell_type, "_", Time),
                   CellType_Time = factor(CellType_Time, levels = unique(CellType_Time))) %>%
            select(CellType_Time, Drug)

df2 = df[,rownames(top_annot)]

#zero out non sig
binary_filt = log2FC.all.binary[rownames(df2),colnames(df2)]
df2_zeroed = df2 * abs(binary_filt) 
df2_zeroed_filt = df2_zeroed %>%
                  mutate_all(~ifelse(abs(.) <1 , 0, .)) %>%
                  filter(rowSums(.!=0) !=  0)  %>%
                  as.matrix()
  
p<-Heatmap(df2_zeroed_filt,
        column_title = "Glu Drug DC-T RNA-seq",
        name = "LFC \n(Drug/NO Drug)",
        show_row_names = F,
        row_title_gp = gpar(fontsize = 6),
        show_column_names=F,
        cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = top_annot,
                         col = list(
                                CellType_Time = c("Tcell_4h" = "#FBE0E0",
                                        "Tcell_24h" = "#F08686",
                                         "Tcell_72h" = "#ED6868", 
                                        "BMDC_4h" = "#F2EEF5",
                                        "BMDC_24h" = "#C0ACD0",
                                        "BMDC_72h" =  "#9989A6"),
                                Drug = c(
                                "DMSO" = "grey",
                                "AAA" = "#A6DFF7",
                                "Erastin" = "#B9CEE6",
                                "Cerestat" = "#8C9FCA", 
                                "D-AP5" = "#6D91CB",
                                "AAA+Cerestat" = "#5BABA4","AAA+D-AP5" = "#68C3A5",
                                "Erastin+Cerestat" ="#539C84",
                                "Erastin+D-AP5" ="#3E7563",
                                "AAA+Erastin" ="#9E7FB8", "D-AP5+Cerestat" = "#5E4C6E")),
                          border = T,
                          show_legend = F,
                          annotation_name_gp = gpar(fontsize = 8)),
        column_split = top_annot$CellType_Time,
        column_gap = unit(c(0,0,3,0,0), "mm"),
        border = T,
        row_km = 50,
        col = colorRamp2(c(-4, -1,0,1,4), c("blue","white", "white","white", "red"))
        )

```

```{r}
#save clusters, re-order
#get clusters----
p_drawn <- draw(p) #draw to lock in row clustering
#save plot
do_plot( "complex_heatmap.pdf", 8, 15, p_drawn)
p_drawn #run to show on screen and visually choose clusters

#use LFC to determine order instead of transcript distribution
rd <- row_order(p_drawn) #get row order & save

#clusters
#1: BMDC up 4h - 30
#2: BMDC up 24h; shared - 29
#3: BMDC up 24h; Erastin + AAA - 35
#4: BMDC up 72h; shared - 41
#5: BMDC up 72h; partial share - 47
#6: BMDC up 72h; AAA driven - 40
#7: BMDC up 72h; Erastin  + Cerestat - 38
#8: BMDC up 72h; cerestat driven - 39
#9: BMDC + T up 72h; shared; 50,48
#10: T up 24h; shared - 49
#11: Tup 24h; Erastin+Cerestat -36
#12: T up 72h; shared - 46
#13: T up 72h; AAA+Erastin - 45
#14: T up 72h; Erastin driven - 44
#15; T up 72h; erastin + ceresat - 42 
#16; Tup 72h; Erastin + NMDA inhibitor - 43
#17; BMDC down 4h; 31,32,34
#18: BMDC down 24h; 33,17
#19; BMDC down 72h; partially shared  - 13,14,15
#20: BMDC down 72h; shared - 3,4
#21: BMDC down 72h; erastin +cerestat only - 7,8
#22: BMDC + T down; 2,5, 6
#23: T 4h down; 1, 12
#24: T 24h down; 16
#25: T 72h down; shared - 9,10, 18, 20,
#26: T 72h down; erastin driven in pair- 22,23
#27: T 72h down; erastin + cerestat; 24,25,26
#28: T 72h down; D-AP5 specific - 27
#29: T 72h down; erastin only -21
#30: T 72h down; Erastin + AAA - 19
#31: T 72h down; Erastin + D-AP5 - 11
#32: T 72h down; AAA + cerestat - 28
#33: unassigned - 37

#reorder to make sure order is 1:20 so we dont have issues when assigning clusters
rd_ordered <- rd[order(as.numeric(names(rd)))] # make sure order is 1:20

#get cluster gene lists
cluster_list = list() #initialize empty list
for(i in 1:50){
  cluster_list[[i]] = data.frame(Cluster = i, 
                                 gene = c(rownames(df2_zeroed_filt)[rd_ordered[[i]]]))
}
cluster_df = do.call(bind_rows, cluster_list)

#edit/condense clusters by eye
cluster_df_edit = cluster_df %>%
                  mutate(old_cluster = Cluster, 
                  Cluster = ifelse(old_cluster %in% c(30),1,
                            ifelse(old_cluster %in% c(50,48, 49, 29, 35), 2,
                            ifelse(old_cluster %in% c(41, 47, 39, 40), 3,
                            ifelse(old_cluster %in% c(38), 4, 
                            ifelse(old_cluster %in% c(36,44,46,45), 5,
                            ifelse(old_cluster %in% c(42), 6, 
                            ifelse(old_cluster %in% c(43), 7 , 
                            ifelse(old_cluster %in% c(3,4,13,14,15,2,5,6), 8, 
                              
                             ifelse(old_cluster %in% c(31,32,34), 9,
                             ifelse(old_cluster %in% c(33,17), 10,
                             ifelse(old_cluster %in% c(7,8), 11, 
                             ifelse(old_cluster %in% c(1,12), 12, 
                             ifelse(old_cluster %in% c(16), 13, 
                             ifelse(old_cluster %in% c(9,10,18,20,22,23, 27), 14, 
                             ifelse(old_cluster %in% c(21), 16, 
                             ifelse(old_cluster %in% c(19), 17, 
                             ifelse(old_cluster %in% c(28), 18, 
                             ifelse(old_cluster %in% c(11), 19,
                             ifelse(old_cluster %in% c(24,25,26), 20, 
                             
                             ifelse(old_cluster %in% c(37), 33, 
                                    34
                                    )))))))))))))))))))
                            )) %>%
                  arrange(Cluster)

rownames(cluster_df_edit) = cluster_df_edit$gene

write.csv(as.matrix(cluster_df_edit), 
         path(output_dir,"20241005_50_cluster_list_FDR0.05_LFC1.csv"))

#re-order H_LFC and re-draw heatmap (please make a function for this soon kate!!!!)----
df_reorder = df2_zeroed_filt[cluster_df_edit$gene,]
left_annot = cluster_df_edit %>% select(Cluster)


u <- df_reorder[cluster_df_edit %>% 
                filter(Cluster == "33") %>% 
                select(gene) %>% unlist(),]

p<-Heatmap(df_reorder,
        column_title = "Glu Drug DC-T RNA-seq",
        name = "LFC \n(Drug/NO Drug)",
        show_row_names = F,
        row_title_gp = gpar(fontsize = 12),
        row_names_gp = gpar(fontsize = 4),
        show_column_names=F,
        cluster_rows =F,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = top_annot,
                         col = list(
                                CellType_Time = c("Tcell_4h" = "#FBE0E0",
                                        "Tcell_24h" = "#F08686",
                                         "Tcell_72h" = "#ED6868", 
                                        "BMDC_4h" = "#F2EEF5",
                                        "BMDC_24h" = "#C0ACD0",
                                        "BMDC_72h" =  "#9989A6"),
                                Drug = c(
                                "DMSO" = "grey",
                                "AAA" = "#A6DFF7",
                                "Erastin" = "#B9CEE6",
                                "Cerestat" = "#8C9FCA", 
                                "D-AP5" = "#6D91CB",
                                "AAA+Cerestat" = "#5BABA4","AAA+D-AP5" = "#68C3A5",
                                "Erastin+Cerestat" ="#539C84",
                                "Erastin+D-AP5" ="#3E7563",
                                "AAA+Erastin" ="#9E7FB8",
                                "D-AP5+Cerestat" = "#5E4C6E")),
                          border = T,
                          show_legend = F,
                          annotation_name_gp = gpar(fontsize = 12)),
        column_split = top_annot$CellType_Time,
        column_gap = unit(c(0,0,3,0,0), "mm"),
        #row_km =10,
        row_split = left_annot$Cluster,
        row_gap = unit(0.5, "mm"),
        border = T,
        col = colorRamp2(c(-3, -1,0,1,3),
                         c("blue","white", "white","white", "red"))
        )

p_drawn <- draw(p)
#save plot
do_plot( "complex_heatmap2.pdf", 8, 15, p_drawn)


rd <-row_order(p_drawn)
rd_ordered <- rd[order(as.numeric(names(rd)))]

cluster_list2 = list()
for(i in 1:10){
  cluster_list2[[i]] = data.frame(Cluster_u = i, 
                                 gene = c(rownames(u)[rd_ordered[[i]]]))
}
cluster_df2 = do.call(bind_rows, cluster_list2)
write.csv(as.matrix(cluster_df2), path(output_dir, "unassigned_1.csv"))

p_drawn
cdf2_edit <- cluster_df2 %>%
             mutate(New_Cluster = ifelse(Cluster_u == 1, 9,
                                  ifelse(Cluster_u == 2, 14,
                                  ifelse(Cluster_u == 10, 5,
                                  ifelse(Cluster_u == 9, 1,
                                  ifelse(Cluster_u == 5, 3,
                                  ifelse(Cluster_u == 8, 2,
                                  ifelse(Cluster_u == 3, 8,
                                  ifelse(Cluster_u == 4, 5,
                                                   33)))))))))
  
```

#iteration 2
```{r}
c4 = cluster_df_edit %>%
     full_join(cdf2_edit, by = "gene") %>%
     mutate(Cluster = ifelse(Cluster ==33, New_Cluster, Cluster))

#re-order H_LFC and re-draw heatmap (please make a function for this soon kate!!!!)----
df_reorder = df2_zeroed_filt[c4$gene,]
left_annot = c4 %>% select(Cluster)


u <- df_reorder[c4 %>% 
                filter(Cluster == "33") %>% 
                select(gene) %>% unlist(),]

p<-Heatmap(u,
        column_title = "Glu Drug DC-T RNA-seq",
        name = "LFC \n(Drug/NO Drug)",
        show_row_names = F,
        row_title_gp = gpar(fontsize = 12),
        row_names_gp = gpar(fontsize = 4),
        show_column_names=F,
        cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = top_annot,
                         col = list(
                                CellType_Time = c("Tcell_4h" = "#FBE0E0",
                                        "Tcell_24h" = "#F08686",
                                         "Tcell_72h" = "#ED6868", 
                                        "BMDC_4h" = "#F2EEF5",
                                        "BMDC_24h" = "#C0ACD0",
                                        "BMDC_72h" =  "#9989A6"),
                                Drug = c(
                                "DMSO" = "grey",
                                "AAA" = "#A6DFF7",
                                "Erastin" = "#B9CEE6",
                                "Cerestat" = "#8C9FCA", 
                                "D-AP5" = "#6D91CB",
                                "AAA+Cerestat" = "#5BABA4","AAA+D-AP5" = "#68C3A5",
                                "Erastin+Cerestat" ="#539C84",
                                "Erastin+D-AP5" ="#3E7563",
                                "AAA+Erastin" ="#9E7FB8",
                                "D-AP5+Cerestat" = "#5E4C6E")),
                          border = T,
                          show_legend = F,
                          annotation_name_gp = gpar(fontsize = 12)),
        column_split = top_annot$CellType_Time,
        column_gap = unit(c(0,0,3,0,0), "mm"),
        row_km =10,
        #row_split = left_annot$Cluster,
        row_gap = unit(0.5, "mm"),
        border = T,
        col = colorRamp2(c(-3, 0,3),
                         c("blue","white", "red"))
        )

p_drawn <- draw(p)
p_drawn
#save plot
do_plot( "unassigned_3.pdf", 8, 15, p_drawn)

rd <-row_order(p_drawn)
rd_ordered <- rd[order(as.numeric(names(rd)))]

cluster_list3 = list()
for(i in 1:10){
  cluster_list3[[i]] = data.frame(Cluster_u3 = i, 
                                 gene = c(rownames(u)[rd_ordered[[i]]]))
}
cluster_df3 = do.call(bind_rows, cluster_list3)
write.csv(as.matrix(cluster_df3), path(output_dir, "unassigned_3.csv"))

cdf3_edit <- cluster_df3 %>%
             mutate(New_Cluster2 = ifelse(Cluster_u3 %in% c(1,2), 8,
                                  ifelse(Cluster_u3 == 8, 2,
                                  ifelse(Cluster_u3 == 10, 5,
                                  ifelse(Cluster_u3 == 3, 18,
                                  ifelse(Cluster_u3 == 9, 1,
                                  ifelse(Cluster_u3 == 6, 21,
                                  ifelse(Cluster_u3 == 5, 21,
                                  ifelse(Cluster_u3 == 7, 1,
                                                   33)))))))))
```

#iteration 3
```{r}
c5 = c4 %>%
     full_join(cdf3_edit, by = "gene") %>%
     mutate(Cluster = ifelse(Cluster ==33, New_Cluster2, Cluster))

#re-order H_LFC and re-draw heatmap (please make a function for this soon kate!!!!)----
df_reorder = df2_zeroed_filt[c5$gene,]
left_annot = c5 %>% select(Cluster)


u <- df_reorder[c5 %>% 
                filter(Cluster == "33") %>% 
                select(gene) %>% unlist(),]

p<-Heatmap(df_reorder,
        column_title = "Glu Drug DC-T RNA-seq",
        name = "LFC \n(Drug/NO Drug)",
        show_row_names = F,
        row_title_gp = gpar(fontsize = 12),
        row_names_gp = gpar(fontsize = 4),
        show_column_names=F,
        cluster_rows =F,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = top_annot,
                         col = list(
                                CellType_Time = c("Tcell_4h" = "#FBE0E0",
                                        "Tcell_24h" = "#F08686",
                                         "Tcell_72h" = "#ED6868", 
                                        "BMDC_4h" = "#F2EEF5",
                                        "BMDC_24h" = "#C0ACD0",
                                        "BMDC_72h" =  "#9989A6"),
                                Drug = c(
                                "DMSO" = "grey",
                                "AAA" = "#A6DFF7",
                                "Erastin" = "#B9CEE6",
                                "Cerestat" = "#8C9FCA", 
                                "D-AP5" = "#6D91CB",
                                "AAA+Cerestat" = "#5BABA4","AAA+D-AP5" = "#68C3A5",
                                "Erastin+Cerestat" ="#539C84",
                                "Erastin+D-AP5" ="#3E7563",
                                "AAA+Erastin" ="#9E7FB8",
                                "D-AP5+Cerestat" = "#5E4C6E")),
                          border = T,
                          show_legend = F,
                          annotation_name_gp = gpar(fontsize = 12)),
        column_split = top_annot$CellType_Time,
        column_gap = unit(c(0,0,3,0,0), "mm"),
        #row_km =10,
        row_split = left_annot$Cluster,
        row_gap = unit(0.5, "mm"),
        border = T,
        col = colorRamp2(c(-3, 0,3),
                         c("blue","white", "red"))
        )

p_drawn <- draw(p)
p_drawn
#save plot
do_plot( "complex_heatmp2.pdf", 8, 15, p_drawn)

rd = row_order(p_drawn)
rd_ordered <- rd[order(as.numeric(names(rd)))]

cluster_list4 = list()
for(i in 1:10){
  cluster_list4[[i]] = data.frame(Cluster_u4 = i, 
                                 gene = c(rownames(u)[rd_ordered[[i]]]))
}
cluster_df4 = do.call(bind_rows, cluster_list4)
write.csv(as.matrix(cluster_df4), path(output_dir, "unassigned_4.csv"))

cdf4_edit <- cluster_df4 %>%
             mutate(New_Cluster4 = ifelse(Cluster_u4 %in% c(10,9), 7,
                                  ifelse(Cluster_u4 %in% c(3,5), 10,
                                  ifelse(Cluster_u4 == 4, 8,
                                  ifelse(Cluster_u4 == 6, 33,
                                  ifelse(Cluster_u4 == 8, 5,
                                  ifelse(Cluster_u4 %in% c(1,2), 12,
                                                   33)))))))
```

#another one
```{r}
c6 = c5 %>%
     full_join(cdf4_edit, by = "gene") %>%
     mutate(Cluster = ifelse(Cluster ==33, New_Cluster4, Cluster))

#re-order H_LFC and re-draw heatmap (please make a function for this soon kate!!!!)----
df_reorder = df2_zeroed_filt[c6$gene,]
left_annot = c6 %>% select(Cluster)


u <- df_reorder[c6 %>% 
                filter(Cluster == "33") %>% 
                select(gene) %>% unlist(),]

p<-Heatmap(u,
        column_title = "Glu Drug DC-T RNA-seq",
        name = "LFC \n(Drug/NO Drug)",
        show_row_names = F,
        row_title_gp = gpar(fontsize = 12),
        row_names_gp = gpar(fontsize = 4),
        show_column_names=F,
        cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = top_annot,
                         col = list(
                                CellType_Time = c("Tcell_4h" = "#FBE0E0",
                                        "Tcell_24h" = "#F08686",
                                         "Tcell_72h" = "#ED6868", 
                                        "BMDC_4h" = "#F2EEF5",
                                        "BMDC_24h" = "#C0ACD0",
                                        "BMDC_72h" =  "#9989A6"),
                                Drug = c(
                                "DMSO" = "grey",
                                "AAA" = "#A6DFF7",
                                "Erastin" = "#B9CEE6",
                                "Cerestat" = "#8C9FCA", 
                                "D-AP5" = "#6D91CB",
                                "AAA+Cerestat" = "#5BABA4","AAA+D-AP5" = "#68C3A5",
                                "Erastin+Cerestat" ="#539C84",
                                "Erastin+D-AP5" ="#3E7563",
                                "AAA+Erastin" ="#9E7FB8",
                                "D-AP5+Cerestat" = "#5E4C6E")),
                          border = T,
                          show_legend = F,
                          annotation_name_gp = gpar(fontsize = 12)),
        column_split = top_annot$CellType_Time,
        column_gap = unit(c(0,0,3,0,0), "mm"),
        row_km =10,
        #row_split = left_annot$Cluster,
        row_gap = unit(0.5, "mm"),
        border = T,
        col = colorRamp2(c(-3, 0,3),
                         c("blue","white", "red"))
        )

p_drawn <- draw(p)
p_drawn
#save plot
do_plot( "unassigned_5.pdf", 8, 15, p_drawn)

rd = row_order(p_drawn)
rd_ordered <- rd[order(as.numeric(names(rd)))]

cluster_list5 = list()
for(i in 1:10){
  cluster_list5[[i]] = data.frame(Cluster_u5 = i, 
                                 gene = c(rownames(u)[rd_ordered[[i]]]))
}
cluster_df5 = do.call(bind_rows, cluster_list5)
write.csv(as.matrix(cluster_df5), path(output_dir, "unassigned_5.csv"))

cdf5_edit <- cluster_df5 %>%
             mutate(New_Cluster5 = ifelse(Cluster_u5 %in% c(7,9), 1,
                                  ifelse(Cluster_u5 %in% c(8), 3,
                                  ifelse(Cluster_u5 == 4, 8,
                                  ifelse(Cluster_u5 == 6, 5,
                                  ifelse(Cluster_u5 == 1, 9,
                                  ifelse(Cluster_u5 == 10, 5,
                                  ifelse(Cluster_u5 %in% c(2,3), 10,
                                                   33))))))))
```

#still at it...
```{r}
c7 = c6 %>%
     full_join(cdf5_edit, by = "gene") %>%
     mutate(Cluster = ifelse(Cluster ==33, New_Cluster5, Cluster))

#re-order H_LFC and re-draw heatmap (please make a function for this soon kate!!!!)----
df_reorder = df2_zeroed_filt[c7$gene,]
left_annot = c7 %>% select(Cluster)


u <- df_reorder[c7 %>% 
                filter(Cluster == "33") %>% 
                select(gene) %>% unlist(),]

p<-Heatmap(u,
        column_title = "Glu Drug DC-T RNA-seq",
        name = "LFC \n(Drug/NO Drug)",
        show_row_names = T,
        row_title_gp = gpar(fontsize = 12),
        row_names_gp = gpar(fontsize = 4),
        show_column_names=F,
        cluster_rows =T,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = top_annot,
                         col = list(
                                CellType_Time = c("Tcell_4h" = "#FBE0E0",
                                        "Tcell_24h" = "#F08686",
                                         "Tcell_72h" = "#ED6868", 
                                        "BMDC_4h" = "#F2EEF5",
                                        "BMDC_24h" = "#C0ACD0",
                                        "BMDC_72h" =  "#9989A6"),
                                Drug = c(
                                "DMSO" = "grey",
                                "AAA" = "#A6DFF7",
                                "Erastin" = "#B9CEE6",
                                "Cerestat" = "#8C9FCA", 
                                "D-AP5" = "#6D91CB",
                                "AAA+Cerestat" = "#5BABA4","AAA+D-AP5" = "#68C3A5",
                                "Erastin+Cerestat" ="#539C84",
                                "Erastin+D-AP5" ="#3E7563",
                                "AAA+Erastin" ="#9E7FB8",
                                "D-AP5+Cerestat" = "#5E4C6E")),
                          border = T,
                          show_legend = F,
                          annotation_name_gp = gpar(fontsize = 12)),
        column_split = top_annot$CellType_Time,
        column_gap = unit(c(0,0,3,0,0), "mm"),
        row_km =15,
        #row_split = left_annot$Cluster,
        row_gap = unit(0.5, "mm"),
        border = T,
        col = colorRamp2(c(-3, 0,3),
                         c("blue","white", "red"))
        )

p_drawn <- draw(p)
p_drawn
#save plot
do_plot( "unassigned_6.pdf", 8, 15, p_drawn)

rd = row_order(p_drawn)
rd_ordered <- rd[order(as.numeric(names(rd)))]

cluster_list6 = list()
for(i in 1:15){
  cluster_list6[[i]] = data.frame(Cluster_u6 = i, 
                                 gene = c(rownames(u)[rd_ordered[[i]]]))
}
cluster_df6 = do.call(bind_rows, cluster_list6)
write.csv(as.matrix(cluster_df6), path(output_dir, "unassigned_6.csv"))

cdf6_edit <- cluster_df6 %>%
             mutate(New_Cluster6 = ifelse(Cluster_u6 %in% c(11,15,6) , 1,
                                  ifelse(Cluster_u6 %in% c(1), 10,
                                  ifelse(Cluster_u6 %in% c(12,13,14) |
                                      gene %in% c("Fastk", "Frat2",
                                                  "Gclm", "Ghitm", "Rtn4",
                                                  "P4hb", "Hsbp1", "Sdhb") , 5,
                                  ifelse(gene %in% c("Gm48909", "Gm9844",
                                                     "Rbm3", "Nampt", "Nsa2"), 2,
                                  ifelse(Cluster_u6 %in% c(10,7), 3,
                                  ifelse(Cluster_u6 == 5, 7,
                                  ifelse(gene %in% c("Galns", "Tmem38b",
                                                     "Mir703","Golt1b"), 3,
                                  ifelse(gene %in% c("Taldo1", "Sdhd",
                                                     "Krr1", "Pgd"), 6,
                                  ifelse(gene %in% c("Gm47283", "Slc7a11", "Kpna3",
                                                     "Gm21887", "Ikbke", "Rapgef2",
                                                     "Uba6"), 9,
                                  ifelse(gene %in% c("Fscn1",  "Cd83", "Cpne1"), 12, 
                                  ifelse(gene %in% c("Creb5", "Abcc1", "Cd1d1",
                                                     "Gls"), 11, 
                                  ifelse(gene %in% c("Rnpep"), 10, 
                                  ifelse(gene %in% c("Akap13"), 14,
                                         33
                                         ))))))))))))))
```

#and again...
```{r}
c8 = c7 %>%
     full_join(cdf6_edit, by = "gene") %>%
     mutate(Cluster = ifelse(Cluster ==33, New_Cluster6, Cluster))
write.csv(as.matrix(c8), path(output_dir, "clusterlist_8.csv"))

#re-order H_LFC and re-draw heatmap (please make a function for this soon kate!!!!)----
df_reorder = df2_zeroed_filt[c8$gene,]
cluster_info = c8 %>% 
             mutate(Cluster = ifelse(Cluster >= 2, Cluster + 1,
                                     Cluster),
                    Cluster = ifelse(Cluster == 22, 2, Cluster),
                    Cluster = ifelse(Cluster == 9, 33, Cluster),
                    Cluster = ifelse(Cluster %in% c(10,11,12), Cluster -1, Cluster),
                    Cluster = ifelse(Cluster == 33, 12, Cluster),
                    Cluster = ifelse(Cluster > 16, Cluster  -1, Cluster)) #%>%

left_annot = cluster_info %>% dplyr::select(Cluster)

write.csv(as.matrix(left_annot), fs::path(output_dir, "clusterlist_8_reorder.csv"))

u <- df_reorder[c8 %>% 
                filter(Cluster == "33") %>% 
                select(gene) %>% unlist(),]

p<-Heatmap(df_reorder,
        column_title = "Glu Drug DC-T RNA-seq",
        name = "LFC \n(Drug/NO Drug)",
        show_row_names = F,
        row_title_gp = gpar(fontsize = 10),
        row_names_gp = gpar(fontsize = 4),
        show_column_names=F,
        cluster_rows =F,
        cluster_columns = F,
        top_annotation = HeatmapAnnotation(df = top_annot,
                         col = list(
                                CellType_Time = c("Tcell_4h" = "#FBE0E0",
                                        "Tcell_24h" = "#F08686",
                                         "Tcell_72h" = "#ED6868", 
                                        "BMDC_4h" = "#F2EEF5",
                                        "BMDC_24h" = "#C0ACD0",
                                        "BMDC_72h" =  "#9989A6"),
                                Drug = c(
                                "DMSO" = "grey",
                                "AAA" = "#A6DFF7",
                                "Erastin" = "#B9CEE6",
                                "Cerestat" = "#8C9FCA", 
                                "D-AP5" = "#6D91CB",
                                "AAA+Cerestat" = "#5BABA4","AAA+D-AP5" = "#68C3A5",
                                "Erastin+Cerestat" ="#539C84",
                                "Erastin+D-AP5" ="#3E7563",
                                "AAA+Erastin" ="#9E7FB8",
                                "D-AP5+Cerestat" = "#5E4C6E")),
                          border = T,
                          show_legend = T,
                          annotation_name_gp = gpar(fontsize = 12)),
        column_split = top_annot$CellType_Time,
        column_gap = unit(c(0,0,3,0,0), "mm"),
        #row_km =5,
        row_split = left_annot$Cluster,
        row_gap = unit(0.5, "mm"),
        border = T,
        col = colorRamp2(c(-3, 0,3),
                         c("blue","white", "red"))
        )

p_drawn <- draw(p)
p_drawn
#save plot
do_plot( "heatmap6_legend.pdf", 8, 10, p_drawn)


```


#gene enrichment on clusters
```{r}
for(i in 16:20){

  gene_list = df_reorder %>%
  data.frame() %>%
  rownames_to_column(var = "gene") %>%
  filter(gene %in% (cluster_info %>% 
                   dplyr::filter(Cluster == i) %>%
                   dplyr::select(gene) %>% unlist())) 

GO <-enrichGO(  gene = gene_list$gene,
                OrgDb         = org.Mm.eg.db,
                keyType       = 'SYMBOL',
                ont           = "ALL",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.1,
                qvalueCutoff = 0.1)

p <- dotplot(GO) + 
     ggtitle(str_c("Cluster", as.character(i))) +
     theme(axis.text.y = element_text(size = 10),
           plot.title = element_text(hjust=0.5)) 
#save plots
save_plot(fs::path(output_dir, 
                   str_c("clusterprofiler/ErichGO_Cluster_", as.character(i), ".pdf")),p, base_aspect_ratio =1.6)
}

```
#other go plots
```{r}
groupGO <-groupGO(gene = gene_list$gene,
        OrgDb = org.Mm.eg.db,
        ont = "BP",
        keyType = 'SYMBOL') 

groupGO.df = summary(groupGO) %>%
             arrange(Count) %>%
             filter(Count > .1 * length(gene_list$gene) ) %>%
             mutate(Description= factor(Description,
                          levels = unique(Description))) %>%
             filter(!grepl("interspecies", Description)) 

p1<-ggplot(groupGO.df, aes(y = Description, x = Count)) +
  geom_col() +
  geom_text(mapping = aes(x = Count + 4, label = GeneRatio), size = 2) 

groupGO <-groupGO(gene = gene_list$gene,
        OrgDb = org.Mm.eg.db,
        ont = "MF",
        keyType = 'SYMBOL') 

groupGO.df = summary(groupGO) %>%
             arrange(Count) %>%
             filter(Count > .1 * length(gene_list$gene) ) %>%
             mutate(Description= factor(Description,
                          levels = unique(Description))) %>%
             filter(!grepl("interspecies", Description)) 

p2<-ggplot(groupGO.df, aes(y = Description, x = Count)) +
  geom_col() +
  geom_text(mapping = aes(x = Count + 4, label = GeneRatio), size = 2)

```


